<a href="https://colab.research.google.com/github/athenaap/data_science_hw/blob/main/Homework4V5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/ogut77/DataScience/master/insurance.csv"
df = pd.read_csv(url)


In [2]:
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

Context in Insurance Data
This dataset is often used to predict charges based on the other variables (age, sex, bmi, children, smoker, region). For example:

Input Variables (X): age, sex, bmi, children, smoker, region (features used to make predictions).

Output Variable (y): charges (what you’re trying to predict).

Describtion of variables
1. Age
Description: The age of the individual (the insured person).
Type: Numerical (integer).
Example Values: 19, 45, 62, etc.
Role in Insurance: Age is a key factor in determining insurance charges. Older individuals often have higher medical costs (and thus higher charges) due to increased health risks.
2. Sex
Description: The gender of the individual.
Type: Categorical (text or binary).
Example Values: "male," "female"
Role in Insurance: Gender can influence insurance charges because health risks and medical expenses may differ between males and females (e.g., pregnancy-related costs for females).
3. BMI (Body Mass Index)
Description: A measure of body fat based on height and weight (calculated as weight in kg divided by height in meters squared).
Type: Numerical (float).
Example Values: 25.3, 30.1, 18.5, etc.
Role in Insurance: Higher BMI often correlates with increased health risks (e.g., obesity-related conditions like diabetes or heart disease), leading to higher insurance charges.
4. Children
Description: The number of children (dependents) covered under the individual’s insurance plan.
Type: Numerical (integer).
Example Values: 0, 1, 3, etc.
Role in Insurance: More children can increase insurance costs slightly, as it may reflect additional healthcare needs, though the effect is often less pronounced than other factors like smoking or age.
5. Smoker
Description: Indicates whether the individual smokes tobacco.
Type: Categorical (text or binary).
Example Values: "yes," "no" .
Role in Insurance: Smoking is a major factor in insurance charges. Smokers typically have much higher medical costs due to risks like lung disease or cancer, so their charges are significantly elevated.
6. Region
Description: The geographic region where the individual lives.
Type: Categorical (text).
Example Values: "northeast," "southeast," "southwest," "northwest" (common in U.S.-based datasets).
Role in Insurance: Charges can vary by region due to differences in healthcare costs, lifestyle factors, or local insurance regulations.
7. Charges
Description: The insurance charges (or premiums/costs) billed to the individual, typically in a currency like USD.
Type: Numerical (float).
Example Values: 1684.52, 11234.89, 32050.23, etc.
Role in Insurance: This is usually the target variable (output) in predictive modeling. It represents the amount the insurance company charges, influenced by all the other columns (age, sex, BMI, etc.).



In [3]:
#1. Check if there is null value in dataset df (5 pt)
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

In [4]:
#2. Assign charges to y  and others to X using df. y is output variable and X is input variables (5 pt)

y = df["charges"]
X = df.drop("charges", axis=1)

In [5]:
#3. Use  get_dummies() function from the pandas library to convert categorical variables in a DataFrame (X).
# Drop first drops the first category’s dummy variable to avoid multicollinearity (5 pt)

X = pd.get_dummies(X, drop_first=True)
X.head()

age     bmi  children  sex_male  smoker_yes  region_northwest  \
0   19  27.900         0     False        True             False   
1   18  33.770         1      True       False             False   
2   28  33.000         3      True       False             False   
3   33  22.705         0      True       False              True   
4   32  28.880         0      True       False              True   

   region_southeast  region_southwest  
0             False              True  
1              True             False  
2              True             False  
3             False             False  
4             False             False

In [6]:
#Use following methods for the evaluation on test and train data
def evalmetric(y,ypred):
 from scipy.stats import pearsonr
 import numpy as np
 e = y - ypred
 mse_f = np.mean(e**2)
 rmse_f = np.sqrt(mse_f)
 mae_f = np.mean(abs(e))
 mape_f = 100*np.mean(abs(e/y))
 crl, _ = pearsonr(y, ypred)
 r2_f = crl*crl
 print("MSE:", mse_f)
 print("RMSE:", rmse_f)
 print("MAE:",mae_f)
 print("MAPE:",mape_f)
 print("R-Squared:", round(r2_f, 4))


In [7]:
#4.Get the correlation between X variables and y variables.(5 pt)
X.corrwith(y)


age                 0.299008
bmi                 0.198341
children            0.067998
sex_male            0.057292
smoker_yes          0.787251
region_northwest   -0.039905
region_southeast    0.073982
region_southwest   -0.043210
dtype: float64

In [8]:
#5.Split a dataset into 25%  of data as test data  and 75% of data as training data ( pt)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)


In [14]:
#6. Using Decision Tree and Linear Regression methods, compare the performance results on both test and training data
#to determine which one is more likely to overfit and which is more likely to underfit.
# Do you think that Lasso and Ridge regularization are more likely to improve the results of Linear model test data,
# or would Random Forest or Boosting methods are more likely to improve the results of Decison tree test data?
#Explain your reasoning.(35 pt)

from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge, Lasso
from sklearn.ensemble import RandomForestRegressor

dt_model = DecisionTreeRegressor()
dt_model.fit(X_train, y_train)

y_train_pred_dt = dt_model.predict(X_train)
y_test_pred_dt = dt_model.predict(X_test)

mse_train_dt = mean_squared_error(y_train, y_train_pred_dt)
mse_test_dt = mean_squared_error(y_test, y_test_pred_dt)

r2_train_dt = r2_score(y_train, y_train_pred_dt)
r2_test_dt = r2_score(y_test, y_test_pred_dt)

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

y_train_pred_lr = lr_model.predict(X_train)
y_test_pred_lr = lr_model.predict(X_test)

mse_train_lr = mean_squared_error(y_train, y_train_pred_lr)
mse_test_lr = mean_squared_error(y_test, y_test_pred_lr)

r2_train_lr = r2_score(y_train, y_train_pred_lr)
r2_test_lr = r2_score(y_test, y_test_pred_lr)



In [20]:
print("Linear Regression:")
print("\nTest:\n")
evalmetric(y_test,y_test_pred_lr)
print("\nTrain:\n")
evalmetric(y_train,y_train_pred_lr)
print("\nDecision Tree:")
print("\nTest:\n")
evalmetric(y_test,y_test_pred_dt)
print("\nTrain:\n")
evalmetric(y_train,y_train_pred_dt)

Linear Regression:

Test:

MSE: 35117755.73613632
RMSE: 5926.023602394469
MAE: 4243.654116653137
MAPE: 44.468185116980976
R-Squared: 0.7676

Train:

MSE: 37004502.18409475
RMSE: 6083.132596294014
MAE: 4183.153367011969
MAPE: 42.26867490005659
R-Squared: 0.745

Decision Tree:

Test:

MSE: 37568936.350116394
RMSE: 6129.350401968906
MAE: 2721.2224196835823
MAPE: 29.87015598831537
R-Squared: 0.7762

Train:

MSE: 182648.17106092346
RMSE: 427.37357318969015
MAE: 19.084122482552342
MAPE: 0.47775094523744194
R-Squared: 0.9987


# We can observe that the accuracy on train data in Decision tree model gives 0.99 and on test data 0.77 it is an indicator of overfitting. Linear Regression can slightly benefit from Lasso and Ridge, however from the results of training and test data there is not much difference meaning that there is no overfitting and therefore Lasso and Ridge won't do much difference. Great improvement for Decision tree is Random Forest or Boosting because it prevents model from overfitting.

In [16]:
#7. Explain performance of linear regressin on test data
# using  Root mean squared error, mean absolute error, mean absolute percentage error and R2 metric (10 pt)
evalmetric(y_test, y_test_pred_lr)

MSE: 35117755.73613632
RMSE: 5926.023602394469
MAE: 4243.654116653137
MAPE: 44.468185116980976
R-Squared: 0.7676


# Linear Regression model explains 76.76% of the variance, but the high RMSE (5926.02) and MAPE (44.47%) indicate significant errors, suggesting some underfitting and potential missing complexity in feature relationships.

In [23]:
#8. Use Random Forest and Boosting methods (XGBoost, LightGBM, and CatBoost)
#to obtain the evaluation scores on  test data.
#Which Boosting technique yielded the best performance on the test data based on the R² metric?
#Did you achieve a better result compared to Random Forest on the test data based on the R² metric?
#If there is improvement on Random forest or boosting methods over decison tree, explain  (30 pt)

from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from catboost import CatBoostRegressor
xgb_model = XGBRegressor()
xgb_model.fit(X_train, y_train)
y_test_pred_xgb = xgb_model.predict(X_test)

lgbm_model = LGBMRegressor()
lgbm_model.fit(X_train, y_train)
y_test_pred_lgbm = lgbm_model.predict(X_test)

cat_model = CatBoostRegressor(verbose=0)
cat_model.fit(X_train, y_train)
y_test_pred_cat = cat_model.predict(X_test)

print("XGBoost:")
evalmetric(y_test, y_test_pred_xgb)
print("\nLightGBM:")
evalmetric(y_test, y_test_pred_lgbm)
print("\nCatBoost:")
evalmetric(y_test, y_test_pred_cat)



[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001090 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 319
[LightGBM] [Info] Number of data points in the train set: 1003, number of used features: 8
[LightGBM] [Info] Start training from score 13267.935814
XGBoost:
MSE: 26433443.13176504
RMSE: 5141.346431798293
MAE: 2957.213261792119
MAPE: 34.57266690344694
R-Squared: 0.8301

LightGBM:
MSE: 22005117.949021608
RMSE: 4690.961303296117
MAE: 2700.720352413941
MAPE: 33.101463593940274
R-Squared: 0.8553

CatBoost:
MSE: 21448141.827996742
RMSE: 4631.213861181185
MAE: 2607.3449326649175
MAPE: 32.198823118532935
R-Squared: 0.8582


# CatBoost showed the best performance (R² = 0.8582), outperforming Random Forest and other boosting methods by effectively reducing variance and improving generalization over the overfitting-prone Decision Tree.